Dirty the dependents of uncacheable nodes

stuhood opened the pull request:
### Problem

The rust level `Node::cacheable` flag is currently only used to mark `@goal_rule`s as uncacheable (because they are allowed to operate on `@sideeffecting` types, such as the `Console` and the `Workspace`). But since the implementation of `cacheable` did not allow it to operate deeply in the Graph, we additionally needed to mark their parent `Select` nodes uncacheable, and could not use the flag in more positions.

Via #7350, #8495, #8347, and #8974, it has become clear that we would like to safely allow nodes deeper in the graph to be uncacheable, as this allows for the re-execution of non-deterministic processes, or re-consumption of un-trackable state, such as:
1. a process receiving stdin from a user
2. an intrinsic rule that pokes an un-watched file on the filesystem
3. reading from a stateful process like git

Note that these would all be intrinsic Nodes: it's not clear that we want to expose this facility to `@rule`s directly.

### Solution

Finish adding support for uncacheable nodes. Fixes #6598.

When an uncacheable node completes, it will now keep the value it completed with (in order to correctly compute a `Generation` value), but it will re-compute the value once per `Session`. The accurate `Generation` value for the uncacheable node allows its dependents to "clean" themselves and not re-run unless the uncacheable node produced a different value than it had before. 

### Result

The `Node::cacheable` flag may be safely used deeper in the graph, with the semantics that requests for any of an uncacheable node's dependents will cause it to re-run once per `Session`. The dependents will not re-run unless the value of the uncacheable node changes (regardless of the `Session`).

stuhood commented:
The first commit removes some unnecessary generic parameters that I encountered while adding the `Context` parameter to `cacheable`, but does not change any logic: the second contains the described change.

stuhood commented:
I started adding another testcase and noticed an issue: can hold off on review. Sorry about that!

stuhood commented:
Ok, reviewable!

illicitonion commented on src/rust/engine/graph/src/lib.rs:
```diff
@@ -878,19 +881,28 @@ impl<N: Node> Graph<N> {
+          // independent of matching Generation values. This is to allow for the behaviour that an
+          // uncacheable Node should always have dirty dependents, transitively.
+          if !entry.node().cacheable(context) || !entry.is_clean(context) {
+            complete_as_dirty = true;
```
`complete_as_dirty` is ignored if `!cacheable` - should this maybe be a tri-state enum we can match on, rather than a function call which we'll repeat and a bool?

stuhood replied:
> rather than a function call which we'll repeat and a bool?

We don't repeat the function call... the calls to `cacheable(..)` here are for the dependencies, while the call inside `Entry::complete` is for the node itself. Given that, I don't think the enum makes sense.

stuhood replied:
On re-read, I think I see what you are suggesting... can try that.

stuhood replied:
So, this isn't straightforward: the `complete_as_dirty` boolean comes into play for most of the states of the `Entry::complete(.., result)` parameter, which means that replacing those two parameters with an enum would require a five state enum I think. I tried it and ran into a roadblock.

illicitonion replied:
Interesting... Do you know what the names of those 5 enum variants would be? I'm curious as to whether that enum would be more clear (even if it's more verbose), given that my read of this code was that there would be 3 enum variants, it seems like I mis-understood the behaviour as it's currently written...

stuhood replied:
The relevant properties are:
* node was/wasn't cleaned - It's possible for a node that depends on an uncacheable node to be "cleaned", which means that our result value is `None` and skips cloning its value to complete it.
* complete as dirty/clean - This applies to both "node was/wasn't cleaned" states: a node that was cleaned without re-running uses it's previous value, but stays dirty. It's not possible for the caller of `Entry::complete` to know this though, because the previous value is stored in the Entry (without shenangians).
* (un)cacheable - When a node is uncacheable, the properties above aren't relevant; otherwise they are.

It's possible that the closure that "cleans" a Node ([here](https://github.com/pantsbuild/pants/blob/5f539c1dc6dadab4c33f451fef3103b13fd91c03/src/rust/engine/graph/src/entry.rs#L242-L261)) could clone the previous value, and then `Graph::complete` would be able to compute the final result value to pass to `Entry::complete`... but I think that would require cloning the previous result value in order to clean things, which is not cheap.

In short: I don't see a good way to do this right now.

stuhood replied:
Perhaps it would clarify things to rename "complete_as_dirty" to "has_dirty_dependents", which makes it less verby, and clarifies that not using that value in some codepaths is "a-ok"?

stuhood commented on src/rust/engine/graph/src/entry.rs:
```diff
@@ -59,29 +59,38 @@ impl Generation {
     }
   }
 
+  /// Iff the value is Clean, mark it Dirty.
```
Soooo, possibly? dirty means: "might run again in any session", uncacheable means: "will only run again in a new session". So the gap between them is that if something uncacheable is dirtied _during_ a session, it's unclear whether to re-run it.

I think that because we're asserting that these are not side-effecting operations, re-running them (in or out of a single session) is safe. Not clear... will take a look at what it might do to the data model.

illicitonion replied:
Yeah, I suspect if we're not going to dirty them, we shouldn't allow deps; I'm imagining something like (I'm aware we don't support non-cacheable `@rule`s _now_, but we should make the framework force us to think about it if we add them):
```python
@rule(cacheable=False)
def foo():
  executable = yield Get[Snapshot](some_path_globs)
  do_sideeffecting_thing_with(executable)
```

where the side-effecting rule has a dep, someone edits the snapshotted file, we re-snapshot because that was invalidated, but we don't re-run the rule because it's side-effecting. Erroring out (ideally on the first `get` from the side-effecting rule, rather than on invalidation), rather than silently not running the rule, feels safer/less surprising.

stuhood replied:
Just ftr: this is _not_ supposed to represent side-effecting things. Those are still allowed only at the root.

stuhood replied:
Oh, apologies. I just realized that your example is still relevant in a `@goal_rule` (marked uncacheable), meaning that sideeffects are in fact relevant here.

But I'm not sure I understand why your example motivates 1) erroring, or 2) re-running the uncacheable rule. If anything, it feels like that motivates not dirtying it, and keeping the "exactly once per session" semantics, which would have the effect that the rule ran once to completion per session, even if things below it and above it in the graph re-ran.

illicitonion replied:
A specific (admittedly, somewhat contrived) example I can think of is:
* Some checked-in file contains a "git branch name which should be used to fetch golden test files"
* Some uncacheable rule will run `git` to get the contents of a file at that sha. It depends on reading that file, to construct the command line to run. Something like:
```python
git_branch_name = read(some_path_in_git)
golden_value = execute("git", "show", f"{git_branch_name}:golden_file")
use_golden_value()
```

As I believe the current PR stands, if this rule is marked as once-per-session, and the contents of `some_path_in_git` changes, we won't re-run this rule.

It's unclear whether we should.
I can argue that we shouldn't re-run (because in intent, "once-per-session" isn't actually an "exactly-once" guarantee, it's an "at-least once" guarantee - the point of the feature is to say "Yeah, this otherwise would be a cache hit, but we're going to pretend it wasn't because we know it reads external state").
I can argue that it should (because something regular which it depends on was invalidated, and that's how invalidation generally works).

Not allowing deps feels like a safe thing to do, because forcing people to decompose their rules to be one of "uncacehable, but no deps" or "if any of its deps change, it will rerun" seems reasonable. (Except ugh then goal_rules still need to be special).
Re-running feels like a safe thing to do, because we're saying these things _shouldn't_ be side-effecting, so it's _safe_ for us to run these things as often as we want.
Allowing deps, but not re-running, feels like it opens up a chance for stale results and unsurprising behaviour.

stuhood replied:
> Except ugh then goal_rules still need to be special

Right... this was my realization above. And it's less about whether we want side-effecting things deeper in the graph (we definitely do not), and more about whether we want multiple types of classifications of nodes (IMO, we probably do not any time soon).

I think the way I see this is that because cacheable rules are deterministic-ish and not side-effecting, automatically retrying them to attempt to improve the observed transactionality of rules is good and healthy, and that's what we do when we dirty nodes. But, it's not strictly required... it's best effort. Given that, not making an effort for uncacheable nodes (and not attempting to differentiate them from side-effecting nodes) seems reasonable.

stuhood commented:
Rebased and added a commit that uses the `SessionId` in `Node::peek`, which required propagating it quite a few places, but fixed tests which were failing because they could no longer view errors rendered in traces.

stuhood commented:
@illicitonion , @cosmicexplorer : Aside from figuring out what order to land this in relative to #8858, I think it is ready to go.

cosmicexplorer commented on src/rust/engine/graph/src/lib.rs:
```diff
@@ -583,31 +590,37 @@ impl<N: Node> InnerGraph<N> {
-  fn reachable_digest_count(&self, roots: &[N]) -> usize {
+  fn reachable_digest_count(&self, roots: &[N], context: &N::Context) -> usize {
+    // TODO: This is a surprisingly expensive method, because it will clone all reachable values by
+    // calling `peek` on them.
```
Good comment!

cosmicexplorer commented on src/rust/engine/graph/src/node.rs:
```diff
@@ -106,13 +106,26 @@ pub trait NodeContext: Clone + Send + 'static {
+  /// have Session-specific semantics. More than one context object might be associated with a
+  /// single caller "session".
+  ///
+  type SessionId: Clone + Debug + Eq;
```
I like this!

cosmicexplorer commented on src/rust/engine/src/scheduler.rs:
```diff
@@ -267,6 +277,14 @@ impl Scheduler {
   }
 
+  ///
+  /// Return all Digests currently in memory in this Scheduler.
```
Possibly add a note on what this is useful for?

stuhood commented:
Renamed `complete_as_dirty` to `has_dirty_dependencies`, and gambled that that wouldn't break CI.

Thanks for the reviews!
